karapace/kafka_rest_apis/consumer_manager.py

```diff
@@ -495,47 +505,49 @@ async def fetch(self, internal_name: Tuple[str, str], content_type: str, formats
-                            message=f"key deserialization error for format {request_format}: {e}",
-                            sub_code=RESTErrorCodes.HTTP_UNPROCESSABLE_ENTITY.value,
-                            content_type=content_type,
-                        )
-                    try:
-                        value = await self.deserialize(msg.value, request_format) if msg.value else None
-                    except DeserializationError as e:
-                        KarapaceBase.unprocessable_entity(
-                            message=f"value deserialization error for format {request_format}: {e}",
-                            sub_code=RESTErrorCodes.HTTP_UNPROCESSABLE_ENTITY.value,
-                            content_type=content_type,
-                        )
-                    element = {
-                        "topic": tp.topic,
-                        "partition": tp.partition,
-                        "offset": msg.offset,
-                        "timestamp": msg.timestamp,
-                        "key": key,
-                        "value": value,
-                    }
-                    response.append(element)
+            for msg in poll_data:
+                try:
+                    key = await self.deserialize(msg.key(), request_format) if msg.key() else None
+                except DeserializationError as e:
+                    KarapaceBase.unprocessable_entity(
+                        message=f"key deserialization error for format {request_format}: {e}",
+                        sub_code=RESTErrorCodes.HTTP_UNPROCESSABLE_ENTITY.value,
+                        content_type=content_type,
+                    )
+                try:
+                    value = await self.deserialize(msg.value(), request_format) if msg.value() else None
+                except DeserializationError as e:
+                    KarapaceBase.unprocessable_entity(
+                        message=f"value deserialization error for format {request_format}: {e}",
+                        sub_code=RESTErrorCodes.HTTP_UNPROCESSABLE_ENTITY.value,
+                        content_type=content_type,
+                    )
+                element = {
+                    "topic": msg.topic(),
+                    "partition": msg.partition(),
+                    "offset": msg.offset(),
+                    # `confluent_kafka.Message.timestamp()` returns a tuple where the first component is
+                    # the timestamp type, see `karapace.kafka.types.Timestamp`
+                    # In case of the `NOT_AVAILABLE` type whatever the timestamp may be, it cannot be trusted
+                    # and should be ignored according to the confluent-kafka documentation:
+                    # https://docs.confluent.io/platform/current/clients/confluent-kafka-python/html/#confluent_kafka.Message
+                    "timestamp": msg.timestamp()[1] if msg.timestamp()[0] != Timestamp.NOT_AVAILABLE else None,
```

Comment:
interesting, seems we can distinguish between when the timestamp was attached to the record from the Broker or from the Producer. I'm wondering if we can add that information in the element itself (to not lose an important information), but probably its an API that should stay constant. I'm wondering if we can add a property to the returned fetched message itself, like two properties that could be `None` or populated based on this result (like: `def produced_timestamp()` and `def broker_received_timestamp()`.

Reply:
Indeed, that's available in the producer API: https://docs.confluent.io/platform/current/clients/confluent-kafka-python/html/index.html#confluent_kafka.Producer.produce

In this case I think it might have less significance, since when producing via Karapace, we don't have a timestamp option, so it's always going to be from the broker.